### PR TITLE
Adding secretProviderClass CR support for vault

### PR DESCRIPTION
**Pull request by sabkat**

### CODE REVIEW

**Comment by stakater-user** [comment]
@sabkat Yikes! You better fix it before anyone else finds out! [Build](https://github.com/stakater-charts/application/commit/83e09225db4531f880a827d5d3322f95d877aa2d/checks) has Failed!

File: application/templates/secretProviderClass.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+{{- if .Values.secretProviderClass.enabled }}
+apiVersion: secrets-store.csi.x-k8s.io/v1alpha1
+kind: SecretProviderClass
+metadata:
+  name: {{.Values.secretProviderClass.name}}
```
Comment (hanzala1234) [inline application/templates/secretProviderClass.yaml:5]: Add namespace field

File: application/templates/secretProviderClass.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+spec:
+  provider: {{.Values.secretProviderClass.provider}}
+  parameters:
+    vaultAddress: {{.Values.secretProviderClass.parameters.vaultAddress}}
```
Comment (hanzala1234) [inline application/templates/secretProviderClass.yaml:9]: I would say remove take this value from .Values.secretProviderClass.vaultAddress

lets not use .parameter

File: application/templates/secretProviderClass.yaml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    roleName: {{.Values.secretProviderClass.parameters.roleName}}
```
Comment (hanzala1234) [inline application/templates/secretProviderClass.yaml:10]: I would say remove take this value from .Values.secretProviderClass.vaultAddress

lets not use .parameter

File: application/templates/secretProviderClass.yaml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    objects: 
+    {{- if .Values.secretProviderClass.parameters.objects }}
+    {{ toYaml .Values.secretProviderClass.parameters.objects | indent 4 }}
+    {{- end }}
+  secretObjects:
+  {{- if .Values.secretProviderClass.secretObjects }}
```
Comment (hanzala1234) [inline application/templates/secretProviderClass.yaml:16]: this should come line above than this

File: application/templates/secretProviderClass.yaml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  {{- range $value := .Values.secretProviderClass.secretObjects }}
+    - data: {{- toYaml $value.data | nindent 8 }}  
+      secretName: {{$value.secretName}}
+      type: {{$value.type}}
```
Comment (hanzala1234) [inline application/templates/secretProviderClass.yaml:20]: Add space between {{ }}

Comment (hanzala1234) [reply application/templates/secretProviderClass.yaml:20]: should be like {{ $value.type }}

**Comment by stakater-user** [comment]
@sabkat Yikes! You better fix it before anyone else finds out! [Build](https://github.com/stakater-charts/application/commit/65d1bcb13ecc21a9a68edb003675192b53d1ac30/checks) has Failed!

**Comment by stakater-user** [comment]
@sabkat Yikes! You better fix it before anyone else finds out! [Build](https://github.com/stakater-charts/application/commit/882c11e6c0c3d30fc5115ae476c7f5782eb9a626/checks) has Failed!

File: application/values.yaml [hunk 6]
```diff
@@ -287,6 +287,25 @@ route:
...
+    objects: |
+      - objectName: "MONGO_HOST"
+        secretPath: "testing/data/mongoDb"
+        secretKey: "MONGO_HOST"
+  secretObjects:
+    - data:
+      - key: MONGO_HOST
+        objectName: host
+      secretName: secret-mongo-host
+      type: Opaque
```
Comment (rasheedamir) [inline application/values.yaml:307]: @sabkat is this is required just to specify one secret? how will it look like if I have 20 secrets?

File: application/templates/secretProviderClass.yaml (hunk 2 above)
Comment (sabkat) [reply application/templates/secretProviderClass.yaml:9]: So for all .parameters this should be done or just vaultAddress ?

File: application/templates/secretProviderClass.yaml (hunk 4 above)
Comment (sabkat) [reply application/templates/secretProviderClass.yaml:16]: ??

File: application/templates/secretProviderClass.yaml (hunk 2 above)
Comment (hanzala1234) [reply application/templates/secretProviderClass.yaml:9]: yes

File: application/templates/secretProviderClass.yaml (hunk 4 above)
Comment (hanzala1234) [reply application/templates/secretProviderClass.yaml:16]: if condition should come before line "secretObjects". so if secretObjects are not defined in values, this field dont come at all

File: application/values.yaml [hunk 7]
```diff
@@ -287,6 +287,25 @@ route:
     termination: edge
     insecureEdgeTerminationPolicy: Redirect
 
+#Values for SecretProviderClass CR
+secretProviderClass:
+  enabled: true
```
Comment (hanzala1234) [inline application/values.yaml:292]: default value should be false

File: application/values.yaml [hunk 8]
```diff
@@ -287,6 +287,25 @@ route:
...
+  name: mongo-host
+  provider: vault
+  parameters:
+    vaultAddress: "http://vault:8200"
+    roleName: "testing-role"
+    objects: |
+      - objectName: "MONGO_HOST"
```
Comment (hanzala1234) [inline application/values.yaml:299]: comment out these values. see other values example

File: README.md [hunk 9]
```diff
@@ -168,3 +168,11 @@ To uninstall the chart:
 | prometheusRule.enabled | Enable prometheusRule for this app | `false` |
 | prometheusRule.additionalLabels | Kubernetes labels object, these additional labels will be added to PrometheusRule CRD | `{}` |
 | prometheusRule.spec.groups | PrometheusRules in their groups to be added | `[]` |
+| secretProviderClass.enabled | Enables Secret Provider Class Custom Resource  | false |
+| secretProviderClass.name | Name of Secret Provider Class Custom Resource  | "" |
+| secretProviderClass.provider | Provider of Secret Provider Class Custom Resource  | "" |
+| secretProviderClass.parameters | Prameters of Secret Provider Class Custom Resource  | [] |
```
Comment (hanzala1234) [inline README.md:174]: not neeeded

File: application/templates/secretProviderClass.yaml (hunk 1 above)
Comment (sabkat) [reply application/templates/secretProviderClass.yaml:5]: done

**Comment by stakater-user** [comment]
@sabkat Yikes! You better fix it before anyone else finds out! [Build](https://github.com/stakater-charts/application/commit/8b192ff77eaf905e9ae78744e87f37e5b6107ae8/checks) has Failed!

File: README.md (hunk 9 above)
Comment (sabkat) [reply README.md:174]: Not needed in readme ?

**Comment by stakater-user** [comment]
@sabkat Yikes! You better fix it before anyone else finds out! [Build](https://github.com/stakater-charts/application/commit/b9a2f53d30f6b4971dad6e386545231b03f8bf0f/checks) has Failed!

**Comment by stakater-user** [comment]
@sabkat Yikes! You better fix it before anyone else finds out! [Build](https://github.com/stakater-charts/application/commit/2fa157363a7f94b5ca1c536e793ccbabdb77c95a/checks) has Failed!

File: application/templates/secretProviderClass.yaml [hunk 10]
```diff
@@ -0,0 +1,26 @@
+{{- if .Values.secretProviderClass.enabled }}
+apiVersion: secrets-store.csi.x-k8s.io/v1alpha1
+kind: SecretProviderClass
+metadata:
+  name: {{ .Values.secretProviderClass.name }}
+  namespace: {{ .Values.secretProviderClass.namespace }}
```
Comment (hanzala1234) [inline application/templates/secretProviderClass.yaml:6]: namespace shoud be {{ .Release.Namespace }}

File: application/templates/secretProviderClass.yaml [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    {{- end }}
+  {{- if .Values.secretProviderClass.secretObjects }}
+  secretObjects:
+  {{- range $value := .Values.secretProviderClass.secretObjects }}
+    - data: {{- toYaml $value.data | nindent 8 }}  
+      secretName: {{ $value.secretName }}
+      type: {{ $value.type }}
+    {{- end }}
+  {{- end }}
+  {{- end }}
```
Comment (hanzala1234) [inline application/templates/secretProviderClass.yaml:26]: fix the indentation. remove extra spaces

**Comment by stakater-user** [comment]
@sabkat Validation Successful

File: application/values.yaml [hunk 12]
```diff
@@ -287,6 +287,25 @@ route:
     termination: edge
     insecureEdgeTerminationPolicy: Redirect
 
+#Values for SecretProviderClass CR
+secretProviderClass:
+  enabled: false
+  #name: mongo-host
```
Comment (hanzala1234) [inline application/values.yaml:293]: make it like this
name: ""

File: application/values.yaml [hunk 13]
```diff
@@ -287,6 +287,25 @@ route:
...
+  #provider: vault
```
Comment (hanzala1234) [inline application/values.yaml:294]: make it provider:

File: application/values.yaml [hunk 14]
```diff
@@ -287,6 +287,25 @@ route:
...
+  #parameters:
```
Comment (hanzala1234) [inline application/values.yaml:295]: uncomment

**Comment by hanzala1234** [comment]
@sabkat  match your values and comments as per convention. see other values how string is defined and how object values are defined. like which values are commented and which one of those are not

File: application/templates/secretProviderClass.yaml (hunk 2 above)
Comment (sabkat) [reply application/templates/secretProviderClass.yaml]: Done

File: application/templates/secretProviderClass.yaml (hunk 4 above)
Comment (sabkat) [reply application/templates/secretProviderClass.yaml:18]: Done

File: application/templates/secretProviderClass.yaml (hunk 3 above)
Comment (sabkat) [reply application/templates/secretProviderClass.yaml]: Done

File: application/templates/secretProviderClass.yaml (hunk 5 above)
Comment (sabkat) [reply application/templates/secretProviderClass.yaml]: Done

File: application/values.yaml (hunk 6 above)
Comment (sabkat) [reply application/values.yaml]: No sir, for multiple secrets can also be used.

File: application/values.yaml (hunk 8 above)
Comment (sabkat) [reply application/values.yaml]: Done

File: README.md (hunk 9 above)
Comment (sabkat) [reply README.md]: Done

File: application/templates/secretProviderClass.yaml (hunk 11 above)
Comment (sabkat) [reply application/templates/secretProviderClass.yaml:26]: Isn't it supposed to be like this ?

File: application/values.yaml (hunk 7 above)
Comment (sabkat) [reply application/values.yaml]: Done

File: application/templates/secretProviderClass.yaml (hunk 10 above)
Comment (sabkat) [reply application/templates/secretProviderClass.yaml]: resolved

File: application/templates/secretProviderClass.yaml (hunk 11 above)
Comment (sabkat) [reply application/templates/secretProviderClass.yaml:25]: done

File: application/values.yaml (hunk 14 above)
Comment (sabkat) [reply application/values.yaml:295]: Done

File: application/values.yaml (hunk 13 above)
Comment (sabkat) [reply application/values.yaml]: done

File: application/values.yaml (hunk 12 above)
Comment (sabkat) [reply application/values.yaml]: done

**Comment by stakater-user** [comment]
@sabkat Yikes! You better fix it before anyone else finds out! [Build](https://github.com/stakater-charts/application/commit/5b23c3fedd1f926a14914243665f7911b3934c90/checks) has Failed!

**Comment by stakater-user** [comment]
@sabkat Yikes! You better fix it before anyone else finds out! [Build](https://github.com/stakater-charts/application/commit/6365afabf3e24fa6c9675a0e4fecf74f84fae933/checks) has Failed!

**Comment by stakater-user** [comment]
@sabkat Yikes! You better fix it before anyone else finds out! [Build](https://github.com/stakater-charts/application/commit/127a8850cba3ce7b07772c2d1c5ac6b48e139772/checks) has Failed!

**Review by hanzala1234 — DISMISSED** [review]

File: application/values.yaml [hunk 15]
```diff
@@ -287,6 +287,24 @@ route:
     termination: edge
     insecureEdgeTerminationPolicy: Redirect
 
+#Values for SecretProviderClass CR
+secretProviderClass:
+  enabled: false
+  name: #mongo-host
+  provider: #vault
+  vaultAddress: #"http://vault:8200"
```
Comment (hanzala1234) [inline application/values.yaml:295]: remove quotation mark here

File: application/values.yaml [hunk 16]
```diff
@@ -287,6 +287,24 @@ route:
...
+  roleName: #"testing-role"
+  #objects: |
+    #- objectName: "MONGO_HOST"
+      #secretPath: "testing/data/mongoDb"
```
Comment (hanzala1234) [inline application/values.yaml:299]: remove quotes

File: application/values.yaml (hunk 15 above)
Comment (sabkat) [reply application/values.yaml]: Done

File: README.md [hunk 17]
```diff
@@ -168,3 +168,10 @@ To uninstall the chart:
 | prometheusRule.enabled | Enable prometheusRule for this app | `false` |
 | prometheusRule.additionalLabels | Kubernetes labels object, these additional labels will be added to PrometheusRule CRD | `{}` |
 | prometheusRule.spec.groups | PrometheusRules in their groups to be added | `[]` |
+| secretProviderClass.enabled | Enables Secret Provider Class Custom Resource  | false |
+| secretProviderClass.name | Name of Secret Provider Class Custom Resource  | "" |
+| secretProviderClass.provider | Provider of Secret Provider Class Custom Resource  | "" |
+| secretProviderClass.vaultAddress | Address of vault  | "http://vault:8200" |
```
Comment (hanzala1234) [inline README.md:174]: i think its readme

File: README.md [hunk 18]
```diff
@@ -168,3 +168,10 @@ To uninstall the chart:
...
+| secretProviderClass.roleName | Name of the role being referred to in vault   | "" |
+| secretProviderClass.objects | The object created from the secret in vault   | [] |
+| secretProviderClass.secretObjects | This creates the kubernetes secret   | "" |
```
Comment (hanzala1234) [inline README.md:177]: isn't it [] ?

**Review by hanzala1234 — APPROVED** [review]

**Pull request merged**